Doc/library/functions.rst

```diff
@@ -1720,50 +1720,84 @@ are always available.  They are listed here in alphabetical order.
-   This has the effect of dividing the input into n-length chunks.
-
-   :func:`zip` should only be used with unequal length inputs when you don't
-   care about trailing, unmatched values from the longer iterables.  If those
-   values are important, use :func:`itertools.zip_longest` instead.
-
-   :func:`zip` in conjunction with the ``*`` operator can be used to unzip a
-   list::
-
-      >>> x = [1, 2, 3]
-      >>> y = [4, 5, 6]
-      >>> zipped = zip(x, y)
-      >>> list(zipped)
-      [(1, 4), (2, 5), (3, 6)]
-      >>> x2, y2 = zip(*zip(x, y))
-      >>> x == list(x2) and y == list(y2)
-      True
+.. function:: zip(*iterables, strict=False)
+
+   Invert a series of iterables, turning rows into columns, and columns into rows.
+
+   Example::
+
+      >>> for item in zip(('a', 'b', 'c'), (1, 2, 3)):
+      ...     print(item)
+      ...
+      ('a', 1)
+      ('b', 2)
+      ('c', 3)
+
+   More formally: :func:`zip` returns an iterator of tuples, where the *i*-th
+   tuple contains the *i*-th element from each of the argument sequences or
+   iterables.
+
+   :func:`zip` returns a lazy iterable. The elements wouldn't be processed
+   until the iterable is iterated on, e.g. by wrapping in a :class:`tuple`.
+
+   One possible wrinkle is that the iterables passed to :func:`zip` could have
+   different lengths; sometimes by design, and sometimes because of a bug in
+   the code that prepared these iterables.  Python offers three different
+   approaches to dealing with this issue:
+
+   * By default, :func:`zip` has short-circuit behavior.  If any of the
+     iterables are shorter than the rest, :func:`zip` will ignore the remainders
+     of the longer iterables, cutting off the result to the length of the shortest
+     iterable::
+
+        >>> tuple(zip(range(3), ['fee', 'fi', 'fo', 'fum']))
+        ((0, 'fee'), (1, 'fi'), (2, 'fo'))
+
+   * :func:`zip` is often used in cases where the iterables are assumed to be
+     equal.  In such cases, it's recommended to use the ``strict=True`` option.
+     Its output is the same as regular :func:`zip`::
+
+        >>> tuple(zip(('a', 'b', 'c'), (1, 2, 3), strict=True))
+        (('a', 1), ('b', 2), ('c', 3))
+
+     Unlike the default behavior, it asserts that the lengths of iterables is
+     identical, raising a :exc:`ValueError` if they aren't:
+
+        >>> tuple(zip(range(3), ['fee', 'fi', 'fo', 'fum'], strict=True))
+        ValueError: zip() argument 2 is longer than argument 1
+
+     Without the ``strict=True`` argument, any bug that results in iterables of
+     different lengths will be silenced, possibly mainfesting as a hard-to-find
+     bug in another part of the program.
+
+   * Shorter iterables can be padded with a constant value to make all the
+     iterables equal. This is done by :func:`itertools.zip_longest`.
+
+   Special cases: With a single iterable argument, :func:`zip` returns an
+   iterator of 1-tuples.  With no arguments, it returns an empty iterator.
```

Comment:
These aren't special cases. I don't think this paragraph is needed.

Reply:
Funny, this paragraph was already there before I changed the documentation. I added the "Special cases:" heading because I do see these as edge cases. Would "Edge cases: " be more logical? Maybe "special" conveys that they have a behavior that's different than normal cases.

Reply:
I personally wouldn't choose to include it, since it follows naturally from the described behavior. *Maybe* the zero-length case could be considered unspecified without this, but that's such an uncommon case with obvious behavior.

But if we keep it, I don't think we need to add "Edge cases:" or "Special cases:" or anything. It makes it seem more important than it really is.

Reply:
Since this is user documentation, not a formal specification, I think it's nice to include what happens in edge cases even if it follows logically from the rest of the docs. And yes, "Edge cases" is the right heading.